README.md
# ext-jh_socialshareprivacy_ttnews
Adds the heise.de socialshareprivacy plugin to each tt_news item.

The extension 'jh_socialshareprivacy_ttnews' is partly mixture of the extensions
'tweet_and_like' written by Nitin Sewtahalsing and 'socialshareprivacy' written by Peter Linzenkirchner.
So there are, of course, adopted code-lines!

## How to install:

- import extension to extension manager
- install extension
- for possible settings see: 'EXT:jh_socialshareprivacy_ttnews/ext_typoscript_setup.txt'
- add marker to your tt_news template: ###SOCIALPRIVACYSHARE###

Note: This extension requires jQuery!

## Contribute

Hopefully everything works, otherwise contact me: Jonathan Heilmann <mail@jonathan-heilmann.de>
You are welcome to report issues in GitHub or create pull requests: [jonathanheilmann/ext-jh_socialshareprivacy_ttnews](https://github.com/jonathanheilmann/ext-jh_socialshareprivacy_ttnews)

ChangeLog
2012-02-20 Jonathan Heilmann <mail@jonathan-heilmann.de>
    * Initial release.

2012-02-20 Jonathan Heilmann <mail@jonathan-heilmann.de>
    * corrected some bugs evoked by changed extension key

2012-02-24 Jonathan Heilmann <mail@jonathan-heilmann.de>
    * some bugfixes (ext_typoscript_setup.txt) was not used
    * with 'plugin.tt_news.tx_jhsocialshareprivacyttnews_pi1.info_link = off' you are now able to hide the information button

2012-02-25 Jonathan Heilmann <mail@jonathan-heilmann.de>
    * added manuel

2012-03-16 Jonathan Heilmann <mail@jonathan-heilmann.de>
    * bugfix (missing parameter in pi1/class.tx_jhsocialshareprivacyttnews_pi1.php in line 109)

2012-04-05 Jonathan Heilmann <mail@jonathan-heilmann.de>
    * updated manuel
    * updated socialshareprivacy plugin to version 1.4
    * added project on forge.typo3.org (http://forge.typo3.org/projects/extension-jh_socialshareprivacy_ttnews)
    * added extension icon

2012-12-02 Jonathan Heilmann <mail@jonathan-heilmann.de>
    * resolves a javascript error in file "jquery.socialshareprivacy.js" on line 130 (bug #43524)

2013-01-19 Jonathan Heilmann <mail@jonathan-heilmann.de>
    * link for facebook iframe is now forced to be absolute
    * resolved some bugs when reading the typoscript setup settings
    * if 'plugin.tt_news.tx_jhsocialshareprivacyttnews_pi1.socialshareprivacy_path' is empty, the jquery.socialshareprivacy.js won't be included
      Use this, if the extension 'socialshareprivacy' is used. Otherwise you will have a conflict.
    * updated manual
    * set extension-state to beta


2013-01-22 Jonathan Heilmann <mail@jonathan-heilmann.de>
    * set 'plugin.tt_news.tx_jhsocialshareprivacyttnews_pi1.css_path = ' to turn off css of extension

2013-01-22 Jonathan Heilmann <mail@jonathan-heilmann.de>
    * fixed bug for setting the default css_path

2013-04-29 Jonathan Heilmann <mail@jonathan-heilmann.de>
    * jquery was not in EXT dir - now in folder res/
    * moved template.js to folder res/
    * added setup 'plugin.tt_news.tx_jhsocialshareprivacyttnews_pi1.jquery_path' to change path for jQuery
    * updated manual
    * changed extension state to stable

2013-05-03 Jonathan Heilmann <mail@jonathan-heilmann.de>
    * updated jquery.socialshareprivacy.js to work with jQuery version > 1.7.x
    * added jquery-1.6.4.min.js to provide jQuery for jquery.socialshareprivacy.deprecated.js
    * added setup 'plugin.tt_news.tx_jhsocialshareprivacyttnews_pi1.use_socialshareprivacy_deprecated' to use jquery.socialshareprivacy.deprecated.js
    * added setup 'plugin.tt_news.tx_jhsocialshareprivacyttnews_pi1.socialshareprivacy_deprecated_path'
    * added setup 'plugin.tt_news.tx_jhsocialshareprivacyttnews_pi1.jquery_old_path' to change path to an old jQuery version
    * updated manual

2014-08-02 Jonathan Heilmann <mail@jonathan-heilmann.de>
	* Added dependencies.
	* Converted manual to rest documentation.

2015-01-24 Jonathan Heilmann <mail@jonathan-heilmann.de>
	* Rewritten for TYPO3 CMS 6.2 LTS
	* Moved JavaScript Libs to footer and dynamic JavaScript to inline
	* Moved translation to setup and removed locallang.xml
	* Removed deprecated JavaScript in setup
	* Updated documentation

2016-09-23  Jonathan Heilmann <mail@jonathan-heilmann.de>
    * Use https scheme for twitter and facebook (google supports https scheme already)
    * Updated dependencies

Classes/Hooks/TxTtNews.php
<?php

namespace Heilmann\JhSocialshareprivacyTtnews\Hooks;

/***************************************************************
*  Copyright notice
*
*  (c) 2013-2016 Jonathan Heilmann <mail@jonathan-heilmann.de>
*  All rights reserved
*
*  This script is part of the TYPO3 project. The TYPO3 project is
*  free software; you can redistribute it and/or modify
*  it under the terms of the GNU General Public License as published by
*  the Free Software Foundation; either version 2 of the License, or
*  (at your option) any later version.
*
*  The GNU General Public License can be found at
*  http://www.gnu.org/copyleft/gpl.html.
*
*  This script is distributed in the hope that it will be useful,
*  but WITHOUT ANY WARRANTY; without even the implied warranty of
*  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
*  GNU General Public License for more details.
*
*  This copyright notice MUST APPEAR in all copies of the script!
***************************************************************/

use \TYPO3\CMS\Core\Utility\ExtensionManagementUtility;

/**
 * Class TxTtNews
 * @package Heilmann\JhSocialshareprivacyTtnews\Hooks
 */
class TxTtNews extends \TYPO3\CMS\Frontend\Plugin\AbstractPlugin {
	var $prefixId      = 'TxTtNews';        // Same as class name
	var $extKey        = 'jh_socialshareprivacy_ttnews';    // The extension key.
	var $scriptRelPath = 'Classes/Hooks/TxTtNews.php';
	var $pi_checkCHash = true;

	/**
	 * The extraItemMarkerProcessor function from tt_news
	 *
	 * @return array    socialshareprivacy bar
	 */
	public function extraItemMarkerProcessor($markerArray, $row, $conf, &$pObj) {
		$this->conf = $pObj->conf['tx_jhsocialshareprivacyttnews_pi1.']; //required to load LocalLang
		$this->pi_loadLL();

		// get configuration of plugin
		$config = $this->conf;

		// get JavaScript Template
		if(isset($config['template'])) {
			$this->tmpl = $pObj->cObj->fileResource($config['template']);
		} else {
			$this->tmpl = $pObj->cObj->fileResource(ExtensionManagementUtility::siteRelPath($this->extKey).'Resources/Private/Templates/template.js');
		}

		// include CSS if set
		if($config['css_path'] != '') {
			$GLOBALS['TSFE']->pSetup['includeCSS.'][$this->extKey] = $config['css_path'];
		}

		//include JavaScript files
		if($config['use_jquery_from_extension'] == 1) {
			$GLOBALS['TSFE']->pSetup['includeJSFooterlibs.'][$this->extKey.'_1'] = $config['jquery_path'];
		}
		if($config['socialshareprivacy_path'] != '') {
			$GLOBALS['TSFE']->pSetup['includeJSFooterlibs.'][$this->extKey.'_2'] = $config['socialshareprivacy_path'];
		}
		$GLOBALS['TSFE']->pSetup['includeJSFooterlibs.'][$this->extKey.'_3'] = $config['sspInlineFunctions_path'];

		// create link to single-view of tt_news article
		$linkconf = array(
			'parameter' => (!empty($pObj->config['singlePid'])) ? $pObj->config['singlePid'] : $GLOBALS['TSFE']->id,
			'additionalParams' => '&tx_ttnews[tt_news]='.$row['uid'],
			'useCacheHash' => true,
			'forceAbsoluteUrl' => true,
			'returnLast' => 'url',
		);
		$link = $pObj->cObj->typoLink('', $linkconf);

		$markerArray['###SOCIALPRIVACYSHARE###'] = $this->getBar($config, $link, $row['uid'], $pObj);

		return $markerArray;
	}


	private function getBar($config, $link, $newsUid, $pObj) {
		$marker['###NEWSUID###'] = $newsUid;

		// default Settings
		$marker['###INFO_LINK###']=$this->getSetting('info_link','http://www.heise.de/ct/artikel/2-Klicks-fuer-mehr-Datenschutz-1333879.html', $pObj);
		$marker['###TXT_HELP###']=$this->pi_getLL('txt_help','',FALSE);
		$marker['###SETTINGS_PERMA###']=$this->pi_getLL('settings_perma','',FALSE);
		$marker['###COOKIE_EXPIRES###']=$this->getSetting('cookie_expires','356', $pObj);
		$marker['###CSS_PATH###']='';
		$marker['###URI###'] = $link;

		$img_path = ExtensionManagementUtility::siteRelPath($this->extKey).'socialshareprivacy/socialshareprivacy/images/';

		$marker['###FB_STATUS###']=$this->getSettingFB('status','on', $pObj);
		$marker['###FB_DUMMY_IMG###']=$GLOBALS['TSFE']->tmpl->getFileName($this->getSettingFB('dummy_img',$img_path.'dummy_facebook.png', $pObj));
		$marker['###TXT_FB_OFF###']=$this->pi_getLL('txt_fb_off','',FALSE);
		$marker['###TXT_FB_ON###']=$this->pi_getLL('txt_fb_on','',FALSE);
		$marker['###FB_PERMA_OPTION###']=$this->getSettingFB('perma_option','on', $pObj);
		$marker['###FB_DISPLAY_NAME###']=$this->getSettingFB('display_name','Facebook', $pObj);
		$marker['###FB_REFERRER_TRACK###']=$this->getSettingFB('referrer_track','', $pObj);
		$marker['###FB_ACTION###']=$this->getSettingFB('action','recommend', $pObj);
		$marker['###FB_LANGUAGE###']=$this->getSettingFB('language','de_DE', $pObj);
		$marker['###FB_TXT_INFO###']=$this->pi_getLL('fb_txt_info','',FALSE);

		// Twitter settings
		$marker['###TW_STATUS###']=$this->getSettingTW('status','on', $pObj);
		$marker['###TW_DUMMY_IMG###']=$GLOBALS['TSFE']->tmpl->getFileName($this->getSettingTW('dummy_img',$img_path.'dummy_twitter.png', $pObj));
		$marker['###TXT_TWITTER_OFF###']=$this->pi_getLL('txt_twitter_off','',FALSE);;
		$marker['###TXT_TWITTER_ON###']=$this->pi_getLL('txt_twitter_on','',FALSE);;
		$marker['###TW_PERMA_OPTION###']=$this->getSettingTW('perma_option','on', $pObj);
		$marker['###TW_DISPLAY_NAME###']=$this->getSettingTW('display_name','Twitter', $pObj);
		$marker['###TW_REFERRER_TRACK###']=$this->getSettingTW('referrer_track','', $pObj);
		$marker['###TW_LANGUAGE###']=$this->getSettingTW('language','en', $pObj);
		$marker['###TW_TXT_INFO###']=$this->pi_getLL('twitter_txt_info','',FALSE);

		// Google settings
		$marker['###GP_STATUS###']=$this->getSettingGP('status','on', $pObj);
		$marker['###GP_DUMMY_IMG###']=$GLOBALS['TSFE']->tmpl->getFileName($this->getSettingGP('dummy_img',$img_path.'dummy_gplus.png', $pObj));
		$marker['###TXT_GPLUS_OFF###']=$this->pi_getLL('txt_gplus_off','',FALSE);;
		$marker['###TXT_GPLUS_ON###']=$this->pi_getLL('txt_gplus_on','',FALSE);;
		$marker['###GP_PERMA_OPTION###']=$this->getSettingGP('perma_option','on', $pObj);
		$marker['###GP_DISPLAY_NAME###']=$this->getSettingGP('display_name','Google+', $pObj);
		$marker['###GP_REFERRER_TRACK###']=$this->getSettingGP('referrer_track','', $pObj);
		$marker['###GP_LANGUAGE###']=$this->getSettingGP('language','de', $pObj);
		$marker['###GP_TXT_INFO###']=$this->pi_getLL('gplus_txt_info','',FALSE);

		// js in template and output in frontend
		$js_conf=str_replace(array_keys($marker),$marker,$this->tmpl);
		$wrapedJs = '<script type="text/javascript">
	/*<![CDATA[*/
<!--
if (typeof sspInlineFunctions == \'undefined\'){
	window.sspInlineFunctions = new Array();
}
sspInlineFunctions.push(function($) {
'.$js_conf.'
});
// -->
	/*]]>*/
</script>';
		// output of plugin
		$content = $wrapedJs . '<div id="socialshareprivacy_'.$newsUid.'"></div>';
		return $this->pi_wrapInBaseClass($content);
	}


	/**
	 * returns given typoscript-value or default if typoscript empty: default settings
	 *
	 * @param  string       $setting
	 * @param  string       $default value
     * @param  array        $pObj
     * @return string
	 **/
	private function getSetting($setting, $default, $pObj) {
		$lConf = $pObj->conf['tx_jhsocialshareprivacyttnews_pi1.'];
		if ( isset($lConf[$setting]) && $lConf[$setting] == true) {
			return $lConf[$setting];
		} else {
			return $default;
		}
	}

	/**
	 * returns given typoscript-value or default if typoscript empty: facebook settings
	 *
	 * @param  string       $setting
	 * @param  string       $default value
     * @param  array        $pObj
     * @return string
	 **/
	private function getSettingFB($setting, $default, $pObj) {
		$lConf = $pObj->conf['tx_jhsocialshareprivacyttnews_pi1.']['services.']['facebook.'];
		if ( isset($lConf[$setting]) && $lConf[$setting] == true) {
			return $lConf[$setting];
		} else {
			return $default;
		}
	}

	/**
	 * returns given typoscript-value or default if typoscript empty: twitter settings
	 *
	 * @param  string       $setting
	 * @param  string       $default value
     * @param  array        $pObj
     * @return string
	 **/
	private function getSettingTW($setting, $default, $pObj) {
		$lConf = $pObj->conf['tx_jhsocialshareprivacyttnews_pi1.']['services.']['twitter.'];
		if ( isset($lConf[$setting]) && $lConf[$setting] == true) {
			return $lConf[$setting];
		} else {
			return $default;
		}
	}

	/**
	 * returns given typoscript-value or default if typoscript empty: Google+-Settings
	 *
	 * @param  string       $setting
	 * @param  string       $default value
	 * @param  array        $pObj
     * @return string
	 **/
	private function getSettingGP($setting, $default, $pObj) {
		$lConf = $pObj->conf['tx_jhsocialshareprivacyttnews_pi1.']['services.']['gplus.'];
		if ( isset($lConf[$setting]) && $lConf[$setting] == true) {
			return $lConf[$setting];
		} else {
			return $default;
		}
	}

}

Resources/Private/Templates/template.js
(function($){
	if($("#socialshareprivacy_###NEWSUID###").length >= 0){
		$("#socialshareprivacy_###NEWSUID###").socialSharePrivacy({
			services : {
				facebook : {
					'status'	  : '###FB_STATUS###',
					'dummy_img'	  : '###FB_DUMMY_IMG###',
					'txt_fb_off'  : '###TXT_FB_OFF###',
					'txt_fb_on'   : '###TXT_FB_ON###',
					'perma_option': '###FB_PERMA_OPTION###',
					'display_name': '###FB_DISPLAY_NAME###',
					'referrer_track':'###FB_REFERRER_TRACK###',
					'action'	  : '###FB_ACTION###',
					'txt_info'	  : '###FB_TXT_INFO###',
					'language'	  : '###FB_LANGUAGE###'
				},
				twitter : {
					'status'	  : '###TW_STATUS###',
					'dummy_img'	  : '###TW_DUMMY_IMG###',
					'txt_twitter_off'  : '###TXT_TWITTER_OFF###',
					'txt_twitter_on'   : '###TXT_TWITTER_ON###',
					'perma_option': '###TW_PERMA_OPTION###',
					'display_name': '###TW_DISPLAY_NAME###',
					'referrer_track':'###TW_REFERRER_TRACK###',
					'txt_info'	  : '###TW_TXT_INFO###',
					'language'	  : '###TW_LANGUAGE###'
				},
				gplus : {
					'status'	  : '###GP_STATUS###',
					'dummy_img'	  : '###GP_DUMMY_IMG###',
					'txt_gplus_off'  : '###TXT_GPLUS_OFF###',
					'txt_gplus_on'   : '###TXT_GPLUS_ON###',
					'perma_option': '###GP_PERMA_OPTION###',
					'display_name': '###GP_DISPLAY_NAME###',
					'referrer_track':'###GP_REFERRER_TRACK###',
					'txt_info'	  : '###GP_TXT_INFO###',
					'language'	  : '###GP_LANGUAGE###'
				}
			},
		'info_link'         : '###INFO_LINK###',
		'txt_help'          : '###TXT_HELP###',
		'settings_perma'    : '###SETTINGS_PERMA###',
		'cookie_expires'    : '###COOKIE_EXPIRES###',
		'css_path'          : '###CSS_PATH###',
		'uri'				: '###URI###'
		});
	}
})(window.jQuery || window.Zepto);

ext_emconf.php
<?php

/***************************************************************
 * Extension Manager/Repository config file for ext "jh_socialshareprivacy_ttnews".
 *
 * Auto generated 24-01-2015 08:45
 *
 * Manual updates:
 * Only the data in the array - everything else is removed by next
 * writing. "version" and "dependencies" must not be touched!
 ***************************************************************/

$EM_CONF[$_EXTKEY] = array (
	'title' => 'socialshareprivacy for tt_news',
	'description' => 'Adds the heise.de socialshareprivacy plugin to each tt_news item. Marker: ###SOCIALPRIVACYSHARE###',
	'category' => 'plugin',
	'version' => '1.1.0',
	'state' => 'stable',
	'uploadfolder' => true,
	'createDirs' => '',
	'clearcacheonload' => true,
	'author' => 'Jonathan Heilmann',
	'author_email' => 'mail@jonathan-heilmann.de',
	'author_company' => '',
	'constraints' =>
	array (
		'depends' =>
		array (
			'typo3' => '6.2.0-7.6.99',
			'tt_news' => '3.6.0-',
		),
		'conflicts' =>
		array (
		),
		'suggests' =>
		array (
		),
	),
);

